gr-filter/lib/rational_resampler_impl.cc

```diff
@@ -0,0 +1,304 @@
+{
+    return gnuradio::make_block_sptr<rational_resampler_impl<IN_T, OUT_T, TAP_T>>(
+        interpolation, decimation, taps, fractional_bw);
+}
+
+template <class IN_T, class OUT_T, class TAP_T>
+rational_resampler_impl<IN_T, OUT_T, TAP_T>::rational_resampler_impl(
+    unsigned interpolation,
+    unsigned decimation,
+    const std::vector<TAP_T>& taps,
+    float fractional_bw)
+    : block("rational_resampler<IN_T,OUT_T,TAP_T>",
+            io_signature::make(1, 1, sizeof(IN_T)),
+            io_signature::make(1, 1, sizeof(OUT_T))),
+      d_history(1),
+      d_decimation(decimation),
+      d_ctr(0),
+      d_updated(false)
+{
+    if (interpolation == 0) {
+        throw std::out_of_range(
+            "rational_resampler_impl<IN_T,OUT_T,TAP_T>: interpolation must be > 0");
+    }
+    if (decimation == 0) {
+        throw std::out_of_range(
+            "rational_resampler_impl<IN_T,OUT_T,TAP_T>: decimation must be > 0");
+    }
+
+    // If taps are not specified, we need to design them based on fractional_bw
+    if (taps.size() == 0 && fractional_bw <= 0) {
+        fractional_bw = 0.4;
+    }
+
+    auto d = GR_GCD(interpolation, decimation);
+
+    if (taps.size() && (d > 1)) {
+        GR_LOG_INFO(
+            d_logger,
+            boost::format(
+                "Rational resampler has user-provided taps but interpolation (%1%) and "
+                "decimation (%2%) have a GCD of %3%, which increases the complexity of "
+                "the filterbank. Consider reducing these values by the GCD.") %
+                interpolation % decimation % d);
+    }
+
+    std::vector<TAP_T> staps;
+    if (taps.size() == 0) {
+        if (fractional_bw <= 0) {
+            fractional_bw = 0.4;
+        }
+        // If we don't have user-provided taps, reduce the interp and
+        // decim values by the GCD (if there is one) and then define
+        // the taps from these new values.
+        interpolation /= d;
+        decimation /= d;
+        staps = design_filter(interpolation, decimation, fractional_bw);
+    } else {
+        staps = taps;
+    }
+
+    this->set_relative_rate((uint64_t)interpolation, (uint64_t)decimation);
+    this->set_output_multiple(1);
+
+    d_firs.reserve(interpolation);
+    for (unsigned i = 0; i < interpolation; i++) {
+        d_firs.emplace_back(1, std::vector<TAP_T>());
+    }
+
+    set_taps(staps);
+    install_taps(d_new_taps);
+}
+
+template <class IN_T, class OUT_T, class TAP_T>
+void rational_resampler_impl<IN_T, OUT_T, TAP_T>::set_taps(const std::vector<TAP_T>& taps)
+{
+    d_new_taps = taps;
+    d_updated = true;
+
+    // round up length to a multiple of the interpolation factor
+    int n = taps.size() % this->interpolation();
+    if (n > 0) {
+        n = this->interpolation() - n;
+        while (n-- > 0) {
+            d_new_taps.insert(d_new_taps.end(), 0);
+        }
+    }
+
+    assert(d_new_taps.size() % this->interpolation() == 0);
```

Comment:
This is a sidebar, but I'm not a fan of asserts.

1) They are disabled on "Release" builds. So for 99% of user builds, they don't even exist.

2) Because they are disabled on "Release" builds, they don't get tested enough. Problems with asserts tend to show up in official operating system packages, when it's too late to fix them.

Reply:
Removing asserts may be worthy of another PR.

Reply:
How about "this can't happen" unless we want to get into checking every alloc.

Reply:
How do you mean? Isn't it possible for client code to `set_taps()` with a size that's incompatible with the interpolation set?

If so then this can happen. I mean "can't happen" as in it's a bug in this file if it's even possible to use the public member functions such that this assert triggers.

Reply:
The caller would almost always `set_taps()` that aren't the size needed, but then the `if (n > 0)` block then pads them out. The `assert` is only triggered if the padding fails for some inexplicable reason.

Reply:
Ah, I misread it before. This is indeed a "can't happen".